### PR TITLE
v1.10

### LINKED ISSUE #31: Update to Fabric 1.21 please
Please update to 1.21

### PR DESCRIPTION
**Pull request by sa-shiro**
- **Added support for Minecraft 1.20.6** 🐺, **1.21** ⚔️, **1.21.3** 💰 **and 1.21.4** 🪵 (closes #31)
  - **Added the Bogged to the dig up animation**
  - Added additional proxy chest items that hold the actual items to the unrendered chest item in the foot slot, preventing Minecraft from stripping away empty item slots and thus messing up the armor and tools save order. Additionally, HandItems are now also stored in the foot slot (index 4 and 5) instead of a separate chest item in the legs slot
  - Added update function to convert the pre-1.20.5 armor and tools save format to the new post-1.20.6 format
  - Added slot information to the pre-1.20.5 save format, including a user-callable function to add this to already saved entities to prevent Minecraft from throwing away important data when upgrading a world to 1.20.5+
  - Added support for new item component and particle syntax (Thanks @sa-shiro)
  - Copied the contents of the legacy folders to their new names and removed unused code
  - Adjusted exclude tag to account for boat ID renames
  - Updated menu trigger to use advancements instead and added menu message header
- Fixed single layer snow no longer producing snow particles
- Fixed mainhand to offhand handling always clearing new mainhand data before overwriting the old mainhand, even if the new data could not be transferred to the offhand
- Fixed attempting to play the block break animation for unsupported blocks on non-solid blocks like air
- Changed play animation on unsupported blocks setting to off by default to prevent unwanted block updates (#16)
- Removed spawnanimations:welcome tag

### CODE REVIEW

**Comment by Tschipcraft** [comment]
Thank you for the pull request. While updating the particle commands is a good start, it's unfortunately not enough. The main broken part of the data pack is the entity hiding system (located in the ehs folder) as it deals with item data which has been completely overhauled. Additionally, I'd like to use overlays instead of overwriting files in the main data folder to maintain backwards compatibility and not draft a new edition just for 1.20.5+

**Comment by sa-shiro** [comment]
Oh what? This is weird it worked for me when I tested it.

I'll have a closer look again later!

**Comment by sa-shiro** [comment]
What I suspect is that I forgot those changes when I forked the repo as I made modifications a while ago to have a working version for 1.20.5+

So basically started fresh when forking

**Comment by Tschipcraft** [comment]
With overlays, I mean additional folders that load and overwrite certain files in specific versions. They [can be added in the pack.mcmeta file](https://minecraft.wiki/w/Data_pack#pack.mcmeta). I haven't used them in this project yet since it wasn't necessary, but you can take a look [at the dev branch of one of my other projects, dynamiclights](https://github.com/Tschipcraft/dynamiclights/tree/dev) for reference.

It's not like the data pack will not load or appear to work without the necessary changes. However, loot will not be saved and restored, and I believe there may even be a chance that mobs will drop chest items.

**Comment by sa-shiro** [comment]
Ahh those overlays, didn't think you meant Datapack overlays 😅 I'll update the branch when I have time

File: minecraft_1205/data/spawnanimations/functions/internal/entity/ehs/remove_invalid_item/check.mcfunction [hunk 1]
```diff
@@ -0,0 +1,4 @@
+##by Tschipcraft
+
+execute if entity @s[nbt={ArmorItems:[{id:"minecraft:bundle",count:1,components:{"minecraft:custom_data":{TsSaRemove:1}}},{},{},{}]}] run function spawnanimations:internal/entity/ehs/remove_invalid_item/remove
+execute unless entity @s[nbt={Item:{id:"minecraft:chest",tag:{TsSaRemove:1b}}}] run tag @s add ts.sa.i.checked
```
Comment (sa-shiro) [inline minecraft_1205/data/spawnanimations/functions/internal/entity/ehs/remove_invalid_item/check.mcfunction:4]: oops forgot to change this

**Comment by Tschipcraft** [comment]
Thanks for actually taking on the challenge. Bundles have been disabled in some versions or moved behind an experimental flag, so using chests would be better for backwards compatibility, the syntax shouldn't really be that different as well as they are both container items, but I would have to check.

Admittedly, there is literally no documentation on my behalf other than what the functions themselves do, since I never expected anyone to actually contribute to it (here we are) and it was enough for me to remember what the parts do. Don't feel pressured to complete this update, I'll do it when I get the time.

I remove the armor to better hide the mobs (#3, #6). Actually shoving them underground would prevent mob farms from working, but would resolve a lot of issues (especially hiding glowing eyes from spiders and enderman). Another idea I had was a mode that doesn't care at which distance the mobs are from players and just always plays their animation, but didn't get around to implement.

**Comment by sa-shiro** [comment]
ahh okay makes sense, i didnt mean your datapack with doc, i meant the commands in general, beside the wiki which offers just the basic command structure, there is nowhere a detailed list of possible arguments without trial & error or actually looking into the src of minecraft, but yeah tbh this is my first time actually messing with mcfunctions and commands stuff :sweat_smile: 

with the overlay approach it would only be used in versions where bundles exist and it isnt really hidden behind experimental flags, the item exists without, the recipe is just missing to actually get the bundle in survival
It used to be the same, but now its seperate:
Bundle: `minecraft:bundle_contents`
Chests: `minecraft:container`

some better debugging would make stuff so much easier to get to know which function throws a error, its not always visible (like with the latest pr, something throws / spams a error, but idk where it occures)

**Comment by sa-shiro** [comment]
Yeah i figured that out too that empty slots are being stripped off so you cant even store "air" as a empty slot workaround, it has to have a item
> however i didnt find a way to properly insert items into it without the inventory being filled with items to be replaced with..
due to their massive change, this is going to be hard to be implemented

Perhaps we could use a armor stand stored somewhere inaccessible and invisible to the player? would may be a better approach than the containers inside a container

**Comment by Tschipcraft** [comment]
I don't want to deal with entity linking, and having an armor stand for every hidden entity would be a bit much. Containers inside containers sounds better to me lmao. I will come back to this after my finals in approx. 1 month

**Comment by Tschipcraft** [comment]
I've implemented the mentioned idea now. It seems to work fine 🎉
I still have a few things left to do, like adjusting the conflict functions for the new format (now the legs slot is free) and implementing an update function to convert the old to the new format if users decide to update Minecraft with Spawn Animations installed, but after that we can merge this pr

File: .gitattributes
Comment (Tschipcraft) [inline .gitattributes:1]: Any particular reason you added JSON and mcfunction to have LF line endings, btw?

Comment (sa-shiro) [reply .gitattributes:1]: Oh, some plugin may have modified it automatically xD

**Comment by Tschipcraft** [comment]
Not sure why the workflow failed, literally ran without problem on my machine bruh
```
Validating with mecha v0.95.2

INFO   | stats  Analyzed 81 functions
       | ---------------------------------------------------------------------------------
       | Total commands (336 behind execute)                                    |      583
       | ---------------------------------------------------------------------------------
       |        /execute                                                        |      336
       |                 run <subcommand>                                       |      335
       |                 if block <pos> <block> <subcommand>                    |      147
       |                 if score <target> <targetObjective> matches <range>... |       66
       |                 as <targets> <subcommand>                              |       40
       |                 at <targets> <subcommand>                              |       24
       |                 unless block <pos> <block> <subcommand>                |       19
       |                 if data storage <source> <path> <subcommand>           |       19
       |                 if entity <entities> <subcommand>                      |       15
       |                 positioned <pos> <subcommand>                          |       14
       |                 store result score <targets> <objective> <subcommand>  |       14
       |                 if predicate <predicate> <subcommand>                  |       13
       |                 if data entity <source> <path> <subcommand>            |       13
       |                 unless items entity <entities> <slots> <item_predic... |       13
       |                 unless score <target> <targetObjective> matches <ra... |       12
       |                 unless data entity <source> <path> <subcommand>        |        5
       |                 if blocks <start> <end> <destination> all <subcomma... |        4
       |                 if items entity <entities> <slots> <item_predicate>... |        4
       |                 store result entity <target> <path> double <scale> ... |        3
       |                 anchored <anchor> <subcommand>                         |        3
       |                 unless entity <entities> <subcommand>                  |        3
       |                 if score <target> <targetObjective> <= <source> <so... |        2
       |                 unless predicate <predicate> <subcommand>              |        2
       |                 unless data storage <source> <path> <subcommand>       |        2
       |                 if score <target> <targetObjective> > <source> <sou... |        1
       |                 on attacker <subcommand>                               |        1
       |                 if entity <entities>                                   |        1
       |        /data (43 behind execute)                                       |      130
       |              modify entity <target> <targetPath> set from storage <... |       31
       |              merge entity <target> <nbt>                               |       14
       |              modify storage <target> <targetPath> merge value <value>  |       13
       |              modify storage <target> <targetPath> set from storage ... |       13
       |              modify storage <target> <targetPath> set from entity <... |       12
       |              get entity <target> <path>                                |       10
       |              modify entity <target> <targetPath> set from entity <s... |       10
       |              modify entity <target> <targetPath> set value <value>     |        8
       |              modify entity <target> <targetPath> merge value <value>   |        6
       |              modify storage <target> <targetPath> set value <value>    |        6
       |              modify entity <target> <targetPath> merge from storage... |        2
       |              remove storage <target> <path>                            |        2
       |              merge storage <target> <nbt>                              |        1
       |              get entity <target> <path> <scale>                        |        1
       |              modify storage <target> <targetPath> merge from storag... |        1
       |        /scoreboard (34 behind execute)                                 |      107
       |                    players set <targets> <objective> <score>           |       31
       |                    objectives add <objective> <criteria>               |       16
       |                    objectives remove <objective>                       |       16
       |                    players operation <targets> <targetObjective> <o... |       14
       |                    players reset <targets> <objective>                 |       11
       |                    players add <targets> <objective> <score>           |       10
       |                    players remove <targets> <objective> <score>        |        5
       |                    players get <target> <objective>                    |        3
       |                    players enable <targets> <objective>                |        1
       |        /function (71 behind execute)                                   |      101
       |        /particle (87 behind execute)                                   |       88
       |                  <name> <pos> <delta> <speed> <count> normal           |       87
       |                  <name> <pos> <delta> <speed> <count>                  |        1
       |        /playsound (53 behind execute)                                  |       55
       |                   <sound> block <targets> <pos> <volume> <pitch>       |       53
       |                   <sound> hostile <targets> <pos> <volume> <pitch>     |        1
       |                   <sound> master <targets> <pos> <volume> <pitch>      |        1
       |        /tellraw (26 behind execute)                                    |       36
       |        /tag (5 behind execute)                                         |       30
       |             <targets> remove <name>                                    |       17
       |             <targets> add <name>                                       |       13
       |        /schedule (1 behind execute)                                    |        9
       |                  function <function> <time>                            |        6
       |                  clear <function>                                      |        3
       |        /gamerule (4 behind execute)                                    |        6
       |                  doTileDrops <value>                                   |        2
       |                  sendCommandFeedback <value>                           |        2
       |                  doTileDrops                                           |        1
       |                  sendCommandFeedback                                   |        1
       |        /clone (4 behind execute)                                       |        4
       |               <begin> <end> <destination>                              |        2
       |               <begin> <end> <destination> masked                       |        2
       |        /fill (4 behind execute)                                        |        4
       |        /effect                                                         |        3
       |                give <targets> <effect> <seconds> <amplifier> <hideP... |        2
       |                clear <targets> <effect>                                |        1
       |        /datapack                                                       |        3
       |        /kill (1 behind execute)                                        |        2
       |        /say                                                            |        2
       |        /setblock (1 behind execute)                                    |        1
       |        /tp (1 behind execute)                                          |        1
       |        /execute (1 behind execute)                                     |        1
       | ---------------------------------------------------------------------------------
       | Total selectors                                                        |      431
       | ---------------------------------------------------------------------------------
       |        @s                                                              |      343
       |           [type]                                                       |       18
       |           [scores]                                                     |        9
       |           [tag]                                                        |        8
       |           [nbt]                                                        |        7
       |           [predicate]                                                  |        1
       |        @a                                                              |       67
       |           [scores]                                                     |        5
       |           [gamemode]                                                   |        2
       |           [sort]                                                       |        1
       |           [limit]                                                      |        1
       |        @e                                                              |       20
       |           [tag]                                                        |       34
       |           [type]                                                       |       15
       |           [distance]                                                   |        3
       |           [limit]                                                      |        3
       |           [sort]                                                       |        3
       |           [nbt]                                                        |        1
       |           [predicate]                                                  |        1
       |        @r                                                              |        1
       |        @e with missing or inverted type                                |       13
       |        @e with type                                                    |        7
       |           [type=#spawnanimations:can_wear_armor]                       |        3
       |           [type=minecraft:item]                                        |        3
       |           [type=#spawnanimations:dig_up_animation]                     |        1
       | ---------------------------------------------------------------------------------
       | Scoreboard objectives                                                  |       17
       | ---------------------------------------------------------------------------------
       |        ts.sa.settings (dummy)                                          |       89
       |                       $block                                           |       20
       |                       $activation_dist                                 |       13
       |                       $play_unsupport                                  |       11
       |                       $hide_armor                                      |       10
       |                       $version                                         |       10
       |                       $activation_mode                                 |        9
       |                       #lock_distance                                   |        6
       |                       $gamerule.tile_drops                             |        3
       |                       #temp                                            |        3
       |                       $gamerule.send_feedback                          |        2
       |        ts.sa.timer (dummy)                                             |       17
       |        ts.sa.x (dummy)                                                 |       17
       |                $minus                                                  |        4
       |                $global                                                 |        2
       |        ts.sa.e.speed (dummy)                                           |       16
       |        ts.sa.e.y (dummy)                                               |       14
       |                  #temp                                                 |        5
       |                  $offset                                               |        2
       |        ts.sa.distance (dummy)                                          |       11
       |        ts.sa.e.stage (dummy)                                           |       10
       |        ts.sa.y (dummy)                                                 |        9
       |                $global                                                 |        2
       |        ts.sa.z (dummy)                                                 |        9
       |                $global                                                 |        2
       |        ts.sa.e.height (dummy)                                          |        9
       |        tschipcraft.menu (trigger)                                      |        8
       |        ts.sa.count (dummy)                                             |        6
       |                    $global                                             |        4
       |        ts.sa.e.nbt.Inv (dummy)                                         |        6
       |        ts.sa.e.nbt.Sil (dummy)                                         |        6
       |        ts.sa.welcome (trigger)                                         |        5
       |        ts.sa.e.nbt.CrCh (dummy)                                        |        5
       |        tvc_ignore                                                      |        1
       |                   $global                                              |        1

Done!
```
WorksOnMyMachine

**Comment by sarthakariya** [comment]
hi @Tschipcraft are you working on the 1.21 version of the datapack 
Please confirm sir

**Pull request merged**